runtime/src/main/java/org/corfudb/runtime/collections/Table.java

```diff
@@ -169,10 +159,13 @@ public Table(@Nonnull final TableParameters<K, V, M> tableParameters,
      * Delete a record mapped to the specified key.
      *
      * @param key Key.
-     * @return Previously stored Corfu Record.
+     * @return Previously stored Corfu Record. Throws NoSuchElementException if key is not found.
      */
     @Nullable
     CorfuRecord<V, M> deleteRecord(@Nonnull final K key) {
+        if (!corfuTable.containsKey(key)) {
+            throw new NoSuchElementException("deleteRecord failed as key "+key.toString()+" does not exist");
```

Comment:
It's not clear to me that throwing a NoSuchElementException is a better choice than making it a noop.

Reply:
Exactly, we did argue that delete is an idempotent operation, but the main problem appears to be with *some* stream listeners who may do something potentially incorrect when they see a delete operation on a deleted key. 
So a bug was filed asking corfu to just fail the delete if the key does not exist in the hope that some wrong code path (double delete) can be found using this fix.

Reply:
You don't need to execute a delete on a deleted key. You can simply log it and return, rather than throwing the exception. This way none of the subscribers are going to see a double delete.

```
if (!corfuTable.containsKey(key)) {
    log.warn("Deleting a non-existent key.");
    return CorfuRecord.NON_EXISTENT;
}
```

It all boils down to (I) if this behavior is going to cause more noise in form of spurious bugs and (II) if this API ends up being clumsy to program against. Anyways, you have more context to make this decision.

Reply:
From inspection of the consumers, the return value from delete is not used. Also an exception was explicitly requested so that any caller who is not checking for key existence can be flagged and fixed, but I see your point. Let's perhaps quickly sync up.